syntax = "proto3";

package yandex.cloud.audit.compute;

import "google/protobuf/timestamp.proto";
import "google/protobuf/field_mask.proto";
import "google/rpc/status.proto";
import "yandex/cloud/api/tools/options.proto";
import "yandex/cloud/compute/v1/image.proto";
import "yandex/cloud/compute/v1/instance.proto";
import "yandex/cloud/compute/v1/instance_service.proto";
import "yandex/cloud/audit/common.proto";
import "yandex/cloud/validation.proto";

option go_package = "a.yandex-team.ru/cloud/bitbucket/public-api/yandex/cloud/audit/compute;compute";
option java_package = "yandex.cloud.api.audit.compute";
option java_outer_classname = "InstanceEvents";

message CreateInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;

    string zone_id = 3;
    string platform_id = 4;
    repeated string metadata_keys = 5;
    string metadata_serial_port_enable = 6;
    string service_account_id = 7;
    yandex.cloud.compute.v1.NetworkSettings network_settings = 8;
    yandex.cloud.compute.v1.PlacementPolicy placement_policy = 9;
    yandex.cloud.compute.v1.Os os = 10;
    repeated string product_ids = 11;
    yandex.cloud.compute.v1.ResourcesSpec resources_spec = 12;
    yandex.cloud.compute.v1.AttachedDiskSpec boot_disk_spec = 13;
    repeated yandex.cloud.compute.v1.AttachedDiskSpec secondary_disk_specs = 14;
    repeated yandex.cloud.compute.v1.AttachedFilesystemSpec filesystem_specs = 22;
    repeated yandex.cloud.compute.v1.NetworkInterfaceSpec network_interface_specs = 15;
    string hostname = 16;
    yandex.cloud.compute.v1.Resources resources = 17;
    yandex.cloud.compute.v1.AttachedDisk boot_disk = 18;
    repeated yandex.cloud.compute.v1.AttachedDisk secondary_disks = 19;
    repeated yandex.cloud.compute.v1.AttachedFilesystem filesystems = 23;
    repeated yandex.cloud.compute.v1.NetworkInterface network_interfaces = 20;
    string fqdn = 21;
  }
}

message UpdateInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;

    string platform_id = 3;
    repeated string metadata_keys = 4;
    string metadata_serial_port_enable = 5;
    string service_account_id = 6;
    yandex.cloud.compute.v1.NetworkSettings network_settings = 7;
    yandex.cloud.compute.v1.PlacementPolicy placement_policy = 8;
    google.protobuf.FieldMask update_mask = 9;
    yandex.cloud.compute.v1.ResourcesSpec resources_spec = 10;
    yandex.cloud.compute.v1.Resources resources = 11;
  }
}

message DeleteInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
  }
}

message StartInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
  }
}

message StopInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
  }
}

message RestartInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
  }
}

message AttachInstanceDisk {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    yandex.cloud.compute.v1.AttachedDiskSpec attached_disk_spec = 3;
    repeated yandex.cloud.compute.v1.AttachedDisk secondary_disks = 4;
  }
}

message DetachInstanceDisk {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    oneof disk {
      string disk_id = 3;
      string device_name = 4;
    }
    repeated yandex.cloud.compute.v1.AttachedDisk secondary_disks = 5;
  }
}

message UpdateInstanceMetadata {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    repeated string delete = 3 [(yandex.cloud.api.tools.field).lint_skip.all = true];
    repeated string upsert_keys = 4;
    string upsert_serial_port_enable = 5;
    repeated string metadata_keys = 6;
    string metadata_serial_port_enable = 7;
  }
}

message AddInstanceOneToOneNat {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    string network_interface_index = 3;
    yandex.cloud.compute.v1.OneToOneNatSpec one_to_one_nat_spec = 4;
    repeated yandex.cloud.compute.v1.NetworkInterface network_interfaces = 20;
  }
}

message RemoveInstanceOneToOneNat {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    string network_interface_index = 3;
    repeated yandex.cloud.compute.v1.NetworkInterface network_interfaces = 20;
  }
}

message UpdateInstanceNetworkInterface {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    string network_interface_index = 3;
    google.protobuf.FieldMask update_mask = 4;
    string subnet_id = 5;
    yandex.cloud.compute.v1.PrimaryAddressSpec primary_v4_address_spec = 6;
    yandex.cloud.compute.v1.PrimaryAddressSpec primary_v6_address_spec = 7;
    repeated string security_group_ids = 8;
    repeated yandex.cloud.compute.v1.NetworkInterface network_interfaces = 20;
  }
}

message AttachInstanceFilesystem {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    yandex.cloud.compute.v1.AttachedFilesystemSpec attached_filesystem_spec = 3;
    repeated yandex.cloud.compute.v1.AttachedFilesystem filesystems = 4;
  }
}

message DetachInstanceFilesystem {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
    oneof filesystem {
      string filesystem_id = 3;
      string device_name = 4;
    }
    repeated yandex.cloud.compute.v1.AttachedFilesystem filesystems = 5;
  }
}

message GuestStopInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
  }
}

message CrashInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
  }
}

message PreemptInstance {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  message EventDetails {
    string instance_id = 1 [(required) = true];
    string instance_name = 2;
  }
}
